apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -0,0 +1,237 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Eth.EthereumHeightMonitor do
+  @moduledoc """
+  Periodically calls the Ethereum client node to check for Ethereumm's block height. Publishes
+  internal events or raises alarms accordingly.
+
+  When a new block height is received, it publishes an internal event under the topic `"ethereum_new_height"`
+  with the payload `{:ethereum_new_height, height}`. The event is only published when the received
+  block height is higher than the previously published height.
+
+  When the call to the Ethereum client fails or returns an invalid responnse, it raises an
+  `:ethereum_connection_error` alarm. The alarm is cleared once a valid block height is seen.
+
+  When the call to the Ethereum client returns the same block height for longer than
+  `:ethereum_stalled_sync_threshold_ms`, it raises an `:ethereum_stalled_sync` alarm.
+  The alarm is cleared once the block height starts increasing again.
+  """
+  use GenServer
+  require Logger
+  alias OMG.Eth.Event
+
+  @type events_t() :: [Event.t()]
+
+  @type t() :: %__MODULE__{
+          check_interval_ms: pos_integer(),
+          stall_threshold_ms: pos_integer(),
+          tref: reference() | nil,
+          alarm_module: module(),
+          event_bus: module(),
+          ethereum_height: integer(),
+          synced_at: DateTime.t(),
+          connection_alarm_raised: boolean(),
+          stall_alarm_raised: boolean(),
+          events: events_t()
+        }
+
+  defstruct check_interval_ms: 10_000,
+            stall_threshold_ms: 20_000,
+            tref: nil,
+            alarm_module: nil,
+            event_bus: nil,
+            ethereum_height: 0,
+            synced_at: nil,
+            connection_alarm_raised: false,
+            stall_alarm_raised: false,
+            events: []
+
+  #
+  # GenServer APIs
+  #
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @doc """
+  Return current events that are associated with `EthereumHeightMonitor`.
+  """
+  @spec get_events() :: {:ok, events_t()}
+  def get_events() do
+    GenServer.call(__MODULE__, :get_events)
+  end
+
+  #
+  # GenServer behaviors
+  #
+
+  def init(opts) do
+    _ = Logger.info("Starting Ethereum height monitor.")
+    _ = install_alarm_handler()
+
+    state = %__MODULE__{
```

Comment:
maybe we could unify this and pass all the application settings as part of the supervisor worker start, similarly to alarm_module and event_bus.